drivers/i3c/i3c_mcux.c

```diff
@@ -0,0 +1,978 @@
+
+struct mcux_i3c_config {
+	I3C_Type *base;
+
+	const struct device *clock_dev;
+	clock_control_subsys_t clock_subsys;
+
+	uint32_t i3c_pp_scl_hz;
+	uint32_t i3c_od_scl_hz;
+	uint32_t i2c_scl_hz;
+
+	uint8_t clk_div;
+	uint8_t clk_div_slow;
+	uint8_t clk_div_tc;
+
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pincfg;
+#endif
+
+	void (*irq_config_func)(const struct device *dev);
+};
+
+struct mcux_i3c_data {
+	i3c_master_handle_t handle;
+	struct k_sem device_sync_sem;
+	status_t callback_status;
+
+	struct i3c_dev_list device_list;
+};
+
+static int mcux_i3c_dev_register(const struct device *dev,
+				 struct i3c_device_desc *desc)
+{
+	struct mcux_i3c_data *data = dev->data;
+	struct i3c_dev_list *dev_list = &data->device_list;
+	struct i3c_addr_slots *slots = &data->device_list.addr_slots;
+
+	/*
+	 * If there is a static address for the I3C devices, check
+	 * if this address is free, and there is no other devices of
+	 * the same (pre-assigned) address on the bus.
+	 */
+	if (desc->static_addr != 0U) {
+		if (i3c_addr_slots_status_is_free(slots, desc->static_addr)) {
+			/* Mark address slot as I3C device */
+			i3c_addr_slots_status_set(slots, desc->static_addr,
+						  I3C_ADDR_SLOT_STATUS_I3C_DEV);
+		} else {
+			/* Address slot is not free */
+			__ASSERT_NO_MSG(false);
+			return -EINVAL;
+		}
+	}
+
+	i3c_dev_list_target_add(dev_list, desc);
+
+	LOG_DBG("I3C device registered: %p", desc);
+
+	return 0;
+}
+
+static int mcux_i3c_i2c_dev_register(const struct device *dev,
+				     struct i3c_i2c_device_desc *desc)
+{
+	struct mcux_i3c_data *data = dev->data;
+	struct i3c_dev_list *dev_list = &data->device_list;
+	struct i3c_addr_slots *slots = &data->device_list.addr_slots;
+
+	/*
+	 * Check if the I2C device address is free, and there is
+	 * no other devices of the same (pre-assigned) address
+	 * on the bus.
+	 */
+	if (!i3c_addr_slots_status_is_free(slots, desc->addr)) {
+		__ASSERT_NO_MSG(false);
+		return -EINVAL;
+	}
+
+	/* Mark address slot as I2C device */
+	i3c_addr_slots_status_set(slots, desc->addr,
+				  I3C_ADDR_SLOT_STATUS_I2C_DEV);
+
+	i3c_dev_list_i2c_devce_add(dev_list, desc);
+
+	LOG_DBG("I2C device registered: %p", desc);
+
+	return 0;
+}
+
+static void mcux_i3c_isr(const struct device *dev)
+{
+	const struct mcux_i3c_config *config = dev->config;
+	struct mcux_i3c_data *data = dev->data;
+	I3C_Type *base = config->base;
+
+	I3C_MasterTransferHandleIRQ(base, &data->handle);
+}
+
+static void mcux_i3c_ctrl_transfer_callback(I3C_Type *base,
+					    i3c_master_handle_t *handle,
+					    status_t status, void *userData)
+{
+	struct mcux_i3c_data *data = userData;
+
+	ARG_UNUSED(handle);
+	ARG_UNUSED(base);
+
+	data->callback_status = status;
+	k_sem_give(&data->device_sync_sem);
+}
+
+/**
+ * @brief Convert status to printable string.
+ *
+ * @param status Status value
+ *
+ * @return String associated with the @p status.
+ */
+static const char *mcux_i3c_cb_status_str_get(status_t status)
```

Comment:
shouldn't this be optional? are all these strings useful in production/release builds? maybe condition to `CONFIG_LOG=y`?)

Reply:
I changed it so that it only returns error strings if LOG is at debug level. Otherwise it simply returns an empty string so it prints nothing.